pkg/authorization/authconfig.go

```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
-		splitPath := strings.Split(path, ".")
-		extractedHosts, err := extractHosts(target, splitPath)
-
-		if err != nil {
-			return nil, fmt.Errorf("failed to extract hosts at path %s: %w", path, err)
+		for _, v := range in {
+			store[v] = true
 		}
 
-		parsedHosts, errAppend := appendHosts(hosts, extractedHosts...)
+		keys := make([]string, len(store))
 
-		if errAppend != nil {
-			return nil, fmt.Errorf("failed to append hosts %v: %w", extractedHosts, errAppend)
+		i := 0
+
+		for v := range store {
+			keys[i] = v
+			i++
 		}
 
-		hosts = parsedHosts
+		return keys
 	}
 
-	u := unique(hosts)
-	if len(u) == 0 {
-		return []string{"unknown.host.com"}, nil
+	isURI := func(host string) bool {
+		return strings.HasPrefix(host, "http://") || strings.HasPrefix(host, "https://")
 	}
 
-	return u, nil
-}
+	appendHosts := func(hosts []string, foundHosts ...string) ([]string, error) {
+		var errAllParse []error
 
-func extractHosts(target *unstructured.Unstructured, splitPath []string) ([]string, error) {
-	// extracting as string
-	if foundHost, found, err := unstructured.NestedString(target.Object, splitPath...); err == nil && found {
-		return []string{foundHost}, nil
-	}
+		for _, foundHost := range foundHosts {
+			if isURI(foundHost) {
+				parsedURL, errParse := url.Parse(foundHost)
+				if errParse != nil {
+					errAllParse = append(errAllParse, fmt.Errorf("failed to parse URL %s: %w", foundHost, errParse))
+				}
+
+				hosts = append(hosts, parsedURL.Host)
+			} else {
+				hosts = append(hosts, foundHost)
+			}
+		}
 
-	// extracting as slice of strings
-	if foundHosts, found, err := unstructured.NestedStringSlice(target.Object, splitPath...); err == nil && found {
-		return foundHosts, nil
+		return hosts, errors.Join(errAllParse...)
 	}
 
-	return nil, fmt.Errorf("neither string nor slice of strings found at path %v", splitPath)
-}
+	return func(target *unstructured.Unstructured) ([]string, error) {
+		var errAll []error
 
-func unique(in []string) []string {
-	store := map[string]bool{}
+		extractedHosts := []string{}
 
-	for _, v := range in {
-		store[v] = true
-	}
+		for _, extractor := range extractors {
+			extractedDelegateHosts, err := extractor(target)
+			if err != nil {
+				errAll = append(errAll, err)
 
-	keys := make([]string, len(store))
+				continue
+			}
 
-	i := 0
+			extractedHosts, err = appendHosts(extractedHosts, extractedDelegateHosts...)
+			if err != nil {
+				errAll = append(errAll, err)
+			}
+		}
 
-	for v := range store {
-		keys[i] = v
-		i++
+		return unique(extractedHosts), errors.Join(errAll...)
 	}
-
-	return keys
 }
 
-func appendHosts(hosts []string, foundHosts ...string) ([]string, error) {
-	for _, foundHost := range foundHosts {
-		if isURI(foundHost) {
-			parsedURL, errParse := url.Parse(foundHost)
-			if errParse != nil {
-				return nil, fmt.Errorf("failed to parse URL %s: %w", foundHost, errParse)
-			}
+func NewExpressionHostExtractor(paths []string) spi.HostExtractor {
```

Comment:
I wonder if `NewPathExpressionExtractor` wouldn't be more meaningful. Plus little godoc could help.

Reply:
fixed https://github.com/maistra/odh-platform/pull/49/commits/7cf23bd8829904328a2f2c2990ca91070a186ae6